<?xml version='1.0' encoding='utf-8'?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="panda">

  <xacro:include filename="$(find franka_description)/robots/panda.transmission.xacro" />

  <!-- Inertia parameters -->
  <xacro:property name="link0_mass" value="3.0" />  <!-- This mass might be incorrect -->
  <xacro:property name="link1_mass" value="3.0" />
  <xacro:property name="link2_mass" value="3.0" />
  <xacro:property name="link3_mass" value="2.0" />
  <xacro:property name="link4_mass" value="2.0" />
  <xacro:property name="link5_mass" value="2.0" />
  <xacro:property name="link6_mass" value="2.0" />
  <xacro:property name="link7_mass" value="1.0" />
  <xacro:property name="link0_length" value="0.133" />
  <xacro:property name="link1_length" value="0.200" />
  <xacro:property name="link2_length" value="0.316" />
  <xacro:property name="link3_length" value="0.0825" />
  <xacro:property name="link4_length" value="0.0825" />
  <xacro:property name="link5_length" value="0.384" />
  <xacro:property name="link6_length" value="0.10" />
  <xacro:property name="link7_length" value="0.10" />

  <xacro:macro name="cylinder_inertial" params="radius length mass *origin">
    <inertial>
      <mass value="${mass}" />
      <xacro:insert_block name="origin" />
      <inertia ixx="${0.0833333 * mass * (3 * radius * radius + length * length)}" ixy="0.0" ixz="0.0"
        iyy="${0.0833333 * mass * (3 * radius * radius + length * length)}" iyz="0.0"
        izz="${0.5 * mass * radius * radius}" />
    </inertial>
  </xacro:macro>

  <xacro:macro name="panda_arm" params="arm_id:='panda' description_pkg:='franka_description' connected_to:='' xyz:='0 0 0' rpy:='0 0 0'">

    <xacro:unless value="${not connected_to}">
      <joint name="${arm_id}_joint_${connected_to}" type="fixed">
        <parent link="${connected_to}"/>
        <child link="${arm_id}_link0"/>
        <origin rpy="${rpy}" xyz="${xyz}"/>
      </joint>
    </xacro:unless>

    <link name="${arm_id}_link0">
      <visual>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/visual/link0.dae"/>
        </geometry>
      </visual>
      <collision>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/collision/link0.stl"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertial radius="0.06" length="${link0_length}" mass="${link0_mass}">
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      </xacro:cylinder_inertial>
    </link>

    <link name="${arm_id}_link1">
      <visual>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/visual/link1.dae"/>
        </geometry>
      </visual>
      <collision>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/collision/link1.stl"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertial radius="0.06" length="${link1_length}" mass="${link1_mass}">
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      </xacro:cylinder_inertial>
    </link>

    <joint name="${arm_id}_joint1" type="revolute">
      <!-- <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/> -->
      <origin rpy="0 0 0" xyz="0 0 0.333"/>
      <parent link="${arm_id}_link0"/>
      <child link="${arm_id}_link1"/>
      <axis xyz="0 0 1"/>
      <limit effort="87" lower="-2.8973" upper="2.8973" velocity="2.1750"/>
    </joint>

    <link name="${arm_id}_link2">
      <visual>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/visual/link2.dae"/>
        </geometry>
      </visual>
      <collision>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/collision/link2.stl"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertial radius="0.06" length="${link2_length}" mass="${link2_mass}">
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      </xacro:cylinder_inertial>
    </link>

    <joint name="${arm_id}_joint2" type="revolute">
      <!-- <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.7628" soft_upper_limit="1.7628"/> -->
      <origin rpy="${-pi/2} 0 0" xyz="0 0 0"/>
      <parent link="${arm_id}_link1"/>
      <child link="${arm_id}_link2"/>
      <axis xyz="0 0 1"/>
      <limit effort="87" lower="-1.7628" upper="1.7628" velocity="2.1750"/>
    </joint>

    <link name="${arm_id}_link3">
      <visual>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/visual/link3.dae"/>
        </geometry>
      </visual>
      <collision>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/collision/link3.stl"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertial radius="0.06" length="${link3_length}" mass="${link3_mass}">
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      </xacro:cylinder_inertial>
    </link>

    <joint name="${arm_id}_joint3" type="revolute">
      <!-- <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/> -->
      <origin rpy="${pi/2} 0 0" xyz="0 -0.316 0"/>
      <parent link="${arm_id}_link2"/>
      <child link="${arm_id}_link3"/>
      <axis xyz="0 0 1"/>
      <limit effort="87" lower="-2.8973" upper="2.8973" velocity="2.1750"/>
    </joint>

    <link name="${arm_id}_link4">
      <visual>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/visual/link4.dae"/>
        </geometry>
      </visual>
      <collision>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/collision/link4.stl"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertial radius="0.06" length="${link4_length}" mass="${link4_mass}">
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      </xacro:cylinder_inertial>
    </link>

    <joint name="${arm_id}_joint4" type="revolute">
      <!-- <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-3.0718" soft_upper_limit="-0.0698"/> -->
      <origin rpy="${pi/2} 0 0" xyz="0.0825 0 0"/>
      <parent link="${arm_id}_link3"/>
      <child link="${arm_id}_link4"/>
      <axis xyz="0 0 1"/>
      <limit effort="87" lower="-3.0718" upper="-0.0698" velocity="2.1750"/>
    </joint>

    <link name="${arm_id}_link5">
      <visual>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/visual/link5.dae"/>
        </geometry>
      </visual>
      <collision>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/collision/link5.stl"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertial radius="0.06" length="${link5_length}" mass="${link5_mass}">
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      </xacro:cylinder_inertial>
    </link>

    <joint name="${arm_id}_joint5" type="revolute">
      <!-- <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/> -->
      <origin rpy="${-pi/2} 0 0" xyz="-0.0825 0.384 0"/>
      <parent link="${arm_id}_link4"/>
      <child link="${arm_id}_link5"/>
      <axis xyz="0 0 1"/>
      <limit effort="12" lower="-2.8973" upper="2.8973" velocity="2.6100"/>
    </joint>

    <link name="${arm_id}_link6">
      <visual>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/visual/link6.dae"/>
        </geometry>
      </visual>
      <collision>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/collision/link6.stl"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertial radius="0.06" length="${link6_length}" mass="${link6_mass}">
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      </xacro:cylinder_inertial>
    </link>

    <joint name="${arm_id}_joint6" type="revolute">
      <!-- <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-0.0175" soft_upper_limit="3.7525"/> -->
      <origin rpy="${pi/2} 0 0" xyz="0 0 0"/>
      <parent link="${arm_id}_link5"/>
      <child link="${arm_id}_link6"/>
      <axis xyz="0 0 1"/>
      <limit effort="12" lower="-0.0175" upper="3.7525" velocity="2.6100"/>
    </joint>

    <link name="${arm_id}_link7">
      <visual>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/visual/link7.dae"/>
        </geometry>
      </visual>
      <collision>
        <geometry>
          <mesh filename="package://${description_pkg}/meshes/collision/link7.stl"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertial radius="0.06" length="${link7_length}" mass="${link7_mass}">
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      </xacro:cylinder_inertial>
    </link>

    <joint name="${arm_id}_joint7" type="revolute">
      <!-- <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/> -->
      <origin rpy="${pi/2} 0 0" xyz="0.088 0 0"/>
      <parent link="${arm_id}_link6"/>
      <child link="${arm_id}_link7"/>
      <axis xyz="0 0 1"/>
      <limit effort="12" lower="-2.8973" upper="2.8973" velocity="2.6100"/>
    </joint>

    <link name="${arm_id}_link8">
      <xacro:cylinder_inertial radius="0.06" length="0" mass="0">
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      </xacro:cylinder_inertial>
    </link>

    <joint name="${arm_id}_joint8" type="fixed">
      <origin rpy="0 0 0" xyz="0 0 0.107"/>
      <parent link="${arm_id}_link7"/>
      <child link="${arm_id}_link8"/>
      <axis xyz="0 0 0"/>
    </joint>

    <xacro:panda_transmission prefix="${arm_id}"/>
  </xacro:macro>
</robot>
